### PR TITLE
Fix syncing of the second provisioning request

### LINKED ISSUE #1401: Flaky: Provisioning when A workload is using a provision admission check with retry Should retry when ProvisioningRequestConfig has MaxRetries=2, the succeeded if the second Provisioning request succeeds
**What happened**:

Failed "Provisioning admission check suite: [It] Provisioning when A workload is using a provision admission check with retry Should retry when ProvisioningRequestConfig has MaxRetries=2, the succeeded if the second Provisioning request succeeds"

**What you expected to happen**:

No error wouldn't happen

**How to reproduce it (as minimally and precisely as possible)**:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/1400/pull-kueue-test-integration-main/1731815899859521536

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### LINKED ISSUE #1558: Flaky Provisioning Request integration test
**What happened**:

Flaky test:

`Provisioning when A workload is using a provision admission check with retry Should retry when ProvisioningRequestConfig has MaxRetries>o, and every Provisioning request retry fails`

**How to reproduce it (as minimally and precisely as possible)**:

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-kueue-test-integration-main/1743668546505805824

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Kueue version (use `git describe --tags --dirty --always`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind bug
/kind flake

#### Which issue(s) this PR fixes:

Fixes #1558

#### Special notes for your reviewer:

The root cause of the issue is that the returned map by `activeOrLastPRForChecks` could contain
the pr-1, instead of pr-2. This could happen when on the list of PRs passed to the function
we have [pr-2, pr-1]. I wanted to test `TestReconcile` but it appears that the order of PRs returned by the
List is always alphabetical in unit tests, even though it does not need to be from the production client.

The new unit test fails on the main branch consistently.

I assume it also fixes #1401, but will double check.

#### Does this PR introduce a user-facing change?

```release-note
Fix the synchronization of the admission check state based on the second provisioning request
```

### CODE REVIEW

File: pkg/controller/admissionchecks/provisioning/controller_test.go [hunk 1]
```diff
@@ -724,3 +724,133 @@ func TestReconcile(t *testing.T) {
...
+
+			builder = builder.WithLists(
+				&autoscaling.ProvisioningRequestList{Items: tc.requests},
+				&kueue.ProvisioningRequestConfigList{Items: configs},
+				&kueue.AdmissionCheckList{Items: checks},
+			)
+
+			k8sclient := builder.Build()
+			recorder := &utiltesting.EventRecorder{}
+			controller, _ := NewController(k8sclient, recorder)
```
Comment (alculquicondor) [inline pkg/controller/admissionchecks/provisioning/controller_test.go:848]: Do not obscure the error.

Call `t.Fatalf` if there is an error that doesn't allow us to continue with the rest of the test.

Comment (mimowo) [reply pkg/controller/admissionchecks/provisioning/controller_test.go]: Done, also added it to `TestReconcile` which existed before (and from which I copied this bit).

**Review by tenzen-y — COMMENTED** [review]
Thanks!
/lgtm
/approve

**Comment by tenzen-y** [comment]
@mimowo Could you manually create a cherry-pick PR?

**Comment by mimowo** [comment]
@tenzen-y actually I get a conflict because we forgot to cherry-pick https://github.com/kubernetes-sigs/kueue/pull/1351 in the first place. Should we first cherry-pick that one? If we do so, both PRs will cherry-pick without conflicts.

**Comment by tenzen-y** [comment]
> @tenzen-y actually I get a conflict because we forgot to cherry-pick #1351 in the first place. Should we first cherry-pick that one? If we do so, both PRs will cherry-pick without conflicts.

@mimowo Uhm, IIUC #1351 is a new feature, not a hotfix. So, I think we shouldn't cherry-pick #1351 to the release-0.5 branch.

Could this bug happen without #1351?

**Comment by mimowo** [comment]
> @mimowo Uhm, IIUC #1351 is a new feature, not a hotfix. So, I think we shouldn't cherry-pick #1351 to the release-0.5 branch.

It is a hotfix, it does not introduce new API, just makes the provisioning requests retried.

> Could this bug happen without #1351?

No, the test bug is a flaky integration test introduced by #1351.

**Comment by mimowo** [comment]
> It is a hotfix, it does not introduce new API, just makes the provisioning requests retried.

But we may double check with @mwielgus and @alculquicondor

**Comment by tenzen-y** [comment]
IIRC, #1351 has the below limitations. I'm on the fence about whether we could release #1351 as a patch release.

> When `PrivisioningACC` is enabled, and there are existing ProvisioningRequests, they are going to be recreated.
This may cause a job failures for some long-running jobs which were using the ProvisioningRequests.

**Comment by tenzen-y** [comment]
If all agree that #1351 is a hotfix, I'm ok with cherry-picking #1351 to the release-0.5 branch.

**Comment by mimowo** [comment]
Well, I don't think the release note is accurate. The CA (Cluster Autoscaler) will maintain the nodes even if the Provisioning Request is deleted as long as they are in use. I tested this experimentally. We also added later the integration test to verify Kueue would not evict workload with deleted provisioning request: https://github.com/kubernetes-sigs/kueue/pull/1423.

So, preferably I would remove this part from the release note.

**Comment by mimowo** [comment]
FYI I followed with offline discussion with @mwielgus and @alculquicondor and the intention is to cherry-pick the retry of ProvReq, including the fix. I have prepared the PR: https://github.com/kubernetes-sigs/kueue/pull/1595.

**Comment by tenzen-y** [comment]
> Well, I don't think the release note is accurate. The CA (Cluster Autoscaler) will maintain the nodes even if the Provisioning Request is deleted as long as they are in use. I tested this experimentally. We also added later the integration test to verify Kueue would not evict workload with deleted provisioning request: https://github.com/kubernetes-sigs/kueue/pull/1423.

I see. Thank you for sharing your investigations!

> So, preferably I would remove this part from the release note. IIRC I added it before verifying experimentally.

I'm fine with updating.

**Comment by tenzen-y** [comment]
> FYI I followed with offline discussion with @mwielgus and @alculquicondor and the intention is to cherry-pick the retry of ProvReq, including the fix. I have prepared the PR: https://github.com/kubernetes-sigs/kueue/pull/1595.

Thanks!

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Fix the synchronization of the admission check state based on recreated ProvisioningRequest
```

**Pull request merged**